duplicate.py

```diff
@@ -0,0 +1,116 @@
+from nltk.corpus import stopwords
+from nltk.stem.porter import PorterStemmer
+
+from bugbug import bugzilla
+
+logging.basicConfig(
+    format="%(asctime)s : %(levelname)s : %(message)s", level=logging.INFO
+)
+
+nltk.download("stopwords")
+
+
+class SimilarityLsi:
+    def __init__(self):
+
+        self.corpus = []
+
+        for bug in bugzilla.get_bugs():
+            self.corpus.append([bug["id"], bug["summary"]])
+
+        # cleaning the text
+        self.ps = PorterStemmer()
+        texts = [
+            [
+                self.ps.stem(word)
+                for word in summary.lower().split()
+                if word not in set(stopwords.words("english"))
+            ]
+            for bug_id, summary in self.corpus
+        ]
+
+        # Assigning unique integer ids to all words
+        self.dictionary = Dictionary(texts)
+
+        # conversion to bow
+        corpus_final = [self.dictionary.doc2bow(text) for text in texts]
+
+        # initializing and applying the tfidf transformation model on same corpus,resultant corpus is of same dimensions
+        tfidf = models.TfidfModel(corpus_final)
+        corpus_tfidf = tfidf[corpus_final]
+
+        # transform tfidf corpus to latent 300-D space via LATENT SEMANTIC INDEXING
+        self.lsi = models.LsiModel(
+            corpus_tfidf, id2word=self.dictionary, num_topics=300
+        )
+        corpus_lsi = self.lsi[corpus_tfidf]
+
+        # indexing the corpus
+        self.index = similarities.Similarity(
+            output_prefix="simdata.shdat", corpus=corpus_lsi, num_features=300
+        )
+
+    def get_similar_bugs(self, query, duplicate_of, default=10):
+
+        # transforming the query to latent 300-D space
+        for bug_id, summary in self.corpus:
+            if bug_id == query:
+                query_summary = [
+                    self.ps.stem(word)
+                    for word in summary.lower().split()
+                    if word not in set(stopwords.words("english"))
+                ]
+                break
+        vec_bow = self.dictionary.doc2bow(query_summary)
+        vec_lsi = self.lsi[vec_bow]
+
+        # perform a similarity query against the corpus
+        sims = self.index[vec_lsi]
+        sims = sorted(enumerate(sims), key=lambda item: -item[1])
+
+        # bug_id of the k most similar summaries
+        sim_bug_ids = []
+        master_present = False
+        for i, j in enumerate(sims):
+            if duplicate_of == self.corpus[j[0]][0]:
+                master_present = True
+            if i >= 1 and i <= default:  # since i = 0 returns the query itself
+                sim_bug_ids.append(self.corpus[j[0]][0])
+            if i == 2000:
+                break
+
+        if master_present:
+            return sim_bug_ids
+
+        return False
+
+
+def recall_rate():
+    similarity = SimilarityLsi()
+    total_bugs = 0
+    hits = 0
+    for bug in bugzilla.get_bugs():
+        if bug["dupe_of"]:
```

Comment:
Each bug can have many duplicates.
You have a main bug where all bugs are duplicated to (which has "duplicates" as a list of bugs), and you have many duplicate bugs which are duplicated to the main bug (which have "dupe_of" as a single bug). In the end, they are all duplicates of each other.

Reply:
That is why i chose one of the two ways cause then we'll have false accuracy maybe, where we judge one pair of duplicates two times.

Reply:
If we have multiple duplicate bugs, we should measure the recall for all of them, it's OK if we are judging pairs again, as the main bug will be different in each query.

E.g. say you have bug A, B and C that are duplicates of each other. You should search all duplicates of A and see if B and C are caught. You should search all duplicates of B and see if A and C are caught. Same with C.
It's possible, depending on the implementation, that when you search for duplicates of A you find B and C and when you search for duplicates of B you find only A. So we need to measure all possibilities.

Reply:
how do we calculate performance in the case it doesn't find all the duplicates?

Reply:
That's the recall :)
We just need to compute recall considering all bugs in a duplicate group and not only one.

Reply:
duplicate_group = A + duplicates of A
for bug in duplicate_group:
    found_bugs = find_duplicates(bug)
    # Count how many bugs in duplicate_group are in found_bugs (recall)
    # Count how many bugs in found_bugs are not in duplicate_group (precision)

Reply:
what if all the duplicates are not in the range we're training? currently i am using the similarity method to check that where 2000 is the threshold(it doesn't check after that).

Reply:
The method we need to implement is supposed to search for similar bugs in a very big database. If this one you implemented takes too long for 2000 bugs, then we need to switch to something else.

Reply:
No the model works fine and takes fraction of seconds to find similar bugs from the whole database but a lot of time was spent on checking if that duplicate bug is actually present in the database used,so instead i search for them in similar bugs list returned by the model, only in the top 2000 similars. we need to check that, right?

Reply:
The way to solve this is to get all duplicates in a first pass over the whole DB.
Something like:
```
# A map from bug ID to its duplicate IDs.
duplicates = defaultdict(set)
for bug in bugzilla.get_bugs():
    dupes = bug["duplicates"]
    dupes.append(bug["dupe_of"])

    duplicates[bug['id']] |= dupes
    for dupe in dupes:
        duplicates[dupe].add(bug['id'])
```

So when evaluating get_similar_bugs for a given bug, you don't need to iterate through the bugs again, but you can just compare the results with `duplicates[bug_id]`.

Limiting to 2000 is bound to be imprecise anyways.

Reply:
The recall rate is now 9.6%.

Should we first improve this number or get precision numbers too ? Anyways precision is going to be too low for current model which can really make us afraid ;)

Reply:
Let's get the precision number first. Then, after merging this PR, we will work on trying out additional techniques and comparing the results.